Driver won't recover if a volume does not contain a metadata file

### Problem Statement
In a rare case, if a volume does not contain a metadata file, the driver won't start up (keep crashing) with `manager.NewManagerOrDie`. This is because the driver tries to ensure all volumes are accessible. If not, it will return an error that cause a panic.
https://github.com/cert-manager/csi-lib/blob/9811918c72abe4e60bb2c6303346460a804b12b5/storage/filesystem.go#L112-L122

Here is a sample error code we have seen:
```
panic: failed to start manager: listing existing volumes: open csi-data-dir/inmemfs/csi-04b4520253413b4a3120e28f454cc781c11a759627c25a8bbd3f536e8c1c2020/metadata.json: no such file or directory

goroutine 1 [running]:
github.com/cert-manager/csi-lib/manager.NewManagerOrDie({{0x20d1870, 0xc000490450}, 0x0, {0x20cb308, 0xc0004c5b80}, {0x20e4388, 0x3023568}, 0xc000490588, 0x0, {0x7ffdf920c895, ...}, ...})
	/go/pkg/mod/github.com/cert-manager/csi-lib@v0.5.0/manager/manager.go:237 +0xa5
```

This situation could happen when `Pod is created --> NodePublishVolume called --> volume created --> OOM`. Thus following defer function would never be called:
https://github.com/cert-manager/csi-lib/blob/9811918c72abe4e60bb2c6303346460a804b12b5/driver/nodeserver.go#L56-L62

### The Fix
To fix this issue, we may simply log / delete the volume and proceed the rest of jobs.

--- a/storage/filesystem.go
+++ b/storage/filesystem.go
@@ -47,7 +47,7 @@ type Filesystem struct {
 	baseDir string
 
 	// used by the 'read only' methods
-	fs fs.FS
+	fs fs.StatFS
 
 	// FixedFSGroup is an optional field which will set the gid ownership of all
 	// volume's data directories to this value.
@@ -71,7 +71,7 @@ func NewFilesystem(log logr.Logger, baseDir string) (*Filesystem, error) {
 		baseDir: baseDir,
 		// Use the rootfs as the DirFS so that paths passed to both read &
 		// write methods on this struct use a consistent root.
-		fs: os.DirFS("/"),
+		fs: os.DirFS("/").(fs.StatFS),
 	}
 
 	notMnt, err := mount.IsNotMountPoint(mount.New(""), f.tempfsPath())
@@ -110,7 +110,7 @@ func (f *Filesystem) ListVolumes() ([]string, error) {
 
 	var vols []string
 	for _, dir := range dirs {
-		file, err := f.fs.Open(f.metadataPathForVolumeID(dir.Name()))
+		_, err := f.fs.Stat(f.metadataPathForVolumeID(dir.Name()))
 		switch {
 		case errors.Is(err, fs.ErrNotExist):
 			f.log.Info("Directory exists but does not contain a metadata file - deleting directory and its contents", "volume_id", dir.Name())
@@ -123,8 +123,6 @@ func (f *Filesystem) ListVolumes() ([]string, error) {
 			// discovered a volume/directory that does not contain a metadata file
 			return nil, err
 		}
-		// immediately close the file as we just need to verify it exists
-		file.Close()
 		vols = append(vols, dir.Name())
 	}
 

--- a/storage/filesystem.go
+++ b/storage/filesystem.go
@@ -111,9 +111,16 @@ func (f *Filesystem) ListVolumes() ([]string, error) {
 	var vols []string
 	for _, dir := range dirs {
 		file, err := f.fs.Open(f.metadataPathForVolumeID(dir.Name()))
-		if err != nil {
+		switch {
+		case errors.Is(err, fs.ErrNotExist):
+			f.log.Info("Directory exists but does not contain a metadata file - deleting directory and its contents", "volume_id", dir.Name())
+			if err := f.RemoveVolume(dir.Name()); err != nil {
+				return nil, fmt.Errorf("deleting stale volume: %v", err)
+			}
+			// continue to skip this loop iteration
+			continue
+		case err != nil:
 			// discovered a volume/directory that does not contain a metadata file
-			// TODO: log this error to allow startup to continue
 			return nil, err
 		}
 		// immediately close the file as we just need to verify it exists

--- a/storage/filesystem_test.go
+++ b/storage/filesystem_test.go
@@ -118,6 +118,25 @@ func TestFilesystem_ListVolumes(t *testing.T) {
 	}
 }
 
+func TestFilesystem_ListVolumes_CleansUpCorruptVolumes(t *testing.T) {
+	backend := &Filesystem{
+		fs: fstest.MapFS{
+			"inmemfs/fake-volume/metadata.json": &fstest.MapFile{Data: []byte{}},
+			"inmemfs/fake-emptyvolume/nothing":  &fstest.MapFile{Data: []byte{}},
+		},
+	}
+
+	vols, err := backend.ListVolumes()
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	if len(vols) != 1 {
+		t.Errorf("expected 1 volume to be returned but got: %+v", vols)
+	}
+	if vols[0] != "fake-volume" {
+		t.Errorf("expected only entry to be 'fake-volume' but got: %s", vols[0])
+	}
+}
 func Test_fsGroupForMetadata(t *testing.T) {
 	intPtr := func(i int64) *int64 {
 		return &i

